core/sql/optimizer/BindRelExpr.cpp

```diff
@@ -10600,37 +10600,67 @@ Upsert is also converted into merge when TRAF_UPSERT_MODE is set to MERGE and
+  // we dont need to do this transformation.The incoming row will always be 
+  // unique. So first check if we any of the conditions are satisfied to 
+  //even try the transform
+  NABoolean mustTryTransform = FALSE;
   if (isUpsert() && 
-      (NOT getIsTrafLoadPrep()) && 
-      (NOT (getTableDesc()->isIdentityColumnGeneratedAlways() && getTableDesc()->hasIdentityColumnInClusteringKey())) && 
-      (NOT (getTableDesc()->getClusteringIndex()->getNAFileSet()->hasSyskey())) && 
-        // table has secondary indexes or
-        (getTableDesc()->hasSecondaryIndexes() ||
-          // CQD is set to MERGE  
-          ((CmpCommon::getDefault(TRAF_UPSERT_MODE) == DF_MERGE) &&
-            // omitted current default columns with non-aligned row format tables
-            // or omitted default columns with aligned row format tables 
-            (((NOT isAlignedRowFormat) && omittedCurrentDefaultClassCols) ||
-            (isAlignedRowFormat && omittedDefaultCols))) ||
-          // CQD is set to Optimal, for non-aligned row format with omitted 
-          // current columns, it is converted into merge though it is not
-          // optimal for performance - This is done to ensure that when the 
-          // CQD is set to optimal, non-aligned format would behave like 
-          // merge when any column is  omitted 
-          ((CmpCommon::getDefault(TRAF_UPSERT_MODE) == DF_OPTIMAL) &&
-            ((NOT isAlignedRowFormat) && omittedCurrentDefaultClassCols))
-        ) 
-     )
-     return TRUE;
-  else
-     return FALSE;
+      ( NOT getIsTrafLoadPrep()) &&
+      (NOT (getTableDesc()->isIdentityColumnGeneratedAlways() && 
+            getTableDesc()->hasIdentityColumnInClusteringKey()))  && 
+      (NOT(getTableDesc()->getClusteringIndex()->getNAFileSet()->hasSyskey())))
+    {
+      mustTryTransform = TRUE;
+    }
+
+  // Transform upsert to merge in case of special modes and
+  // omitted default columns
+  // Case 1 :  CQD is set to MERGE, omitted current(timestamp) default 
+  //           columns with  non-aligned row format table or omitted 
+  //           default columns with aligned row format tables 
+
+  // Case 2 :  CQD is set to Optimal, for non-aligned row format with omitted 
+  //           current(timestamp) columns, it is converted into merge 
+  //           though it is not optimal for performance. This is done to ensure
+  //           that when the CQD is set to optimal, non-aligned format would 
+  //           behave like merge when any column is  omitted 
+  if (isUpsert()  &&   
+      mustTryTransform &&          
+      ((CmpCommon::getDefault(TRAF_UPSERT_MODE) == DF_MERGE) &&     
+       (((NOT isAlignedRowFormat) && omittedCurrentDefaultClassCols) ||
+        (isAlignedRowFormat && omittedDefaultCols)))
+      ||
+      ((CmpCommon::getDefault(TRAF_UPSERT_MODE) == DF_OPTIMAL) &&
+       ((NOT isAlignedRowFormat) && omittedCurrentDefaultClassCols)))
+    {
+      toMerge = TRUE;
```

Comment:
If we choose the new tree, we won't get merge behaviour. We will get "replace" behaviour. I do think we need to make DF_OPTIMAL the default behaviour so the more efficient one is the default behaviour and if the user wants merge behaviour they can choose the slower method. We can discuss that separately.